geomesa-gs-monitor-elasticsearch/src/main/scala/org/geomesa/gs/monitor/elastic/ElasticRequestDataListener.scala

```diff
@@ -108,33 +120,39 @@ class ElasticRequestDataListener extends RequestDataListener with LazyLogging {
   override def requestStarted(requestData: monitor.RequestData): Unit = {
     val rd = new RequestData(requestData)
 
-    requestsInProgress.synchronized {
-      requestsInProgress.add(rd) // keep track of in-progress request
-    }
+    requestsInProgress.put(rd.internalid, rd) // keep track of in-progress request
   }
 
   override def requestUpdated(requestData: monitor.RequestData): Unit = {
     val rd = new RequestData(requestData)
 
-    requestsInProgress.synchronized {
-      requestsInProgress.add(rd) // add request again to update state
-    }
+    requestsInProgress.put(rd.internalid, rd) // add request again to update state
   }
 
   override def requestCompleted(requestData: monitor.RequestData): Unit = {
     val rd = new RequestData(requestData)
 
-    requestsInProgress.synchronized {
-      requestsInProgress.add(rd) // add request again to update state
-    }
+    requestsInProgress.put(rd.internalid, rd) // add request again to update state
   }
 
   override def requestPostProcessed(requestData: monitor.RequestData): Unit = {
     val rd = new RequestData(requestData)
 
-    requestsInProgress.synchronized {
-      requestsInProgress.remove(rd) // stop tracking finished request
-      writeQueue.add(rd) // submit finished request to be sent to elasticsearch
+    requestsInProgress.compute(rd.internalid, new BiFunction[Long, RequestData, RequestData] {
+      override def apply(id: Long, request: RequestData): RequestData = {
+        writeQueue.add(request)
```

Comment:
i think this should be the `rd` defined above, and not the `request` (which may be null if it timed out and was removed)?